pkg/basinprovider/provider.go

```diff
@@ -132,14 +135,79 @@ func (bp *BasinProvider) Push(ctx context.Context, ns string, rel string, tx bas
 	return err
 }
 
+// Upload uploads a file to th server.
+func (bp *BasinProvider) Upload(
+	ctx context.Context, ns string, rel string, r io.Reader, signer *app.Signer, progress io.Writer,
+) error {
+	uploadFuture, uploadRelease := bp.p.Upload(ctx, func(p Publications_upload_Params) error {
```

Comment:
calls Upload API method with `ns` and `rel` and get a callback

Reply:
do you need to call `Struct` on `uploadFuture` to catch the error?